### PR TITLE
wifi: Add ext flash support for firmware patches

### PR DESCRIPTION
**Pull request by krish2718**
Pull in prerequisites to add external flash support for nRF70 firmware patches

### CODE REVIEW

File: subsys/net/lib/Kconfig [hunk 1]
```diff
@@ -47,5 +47,5 @@ rsource "wifi_ready/Kconfig"
 rsource "mqtt_helper/Kconfig"
 rsource "nrf_provisioning/Kconfig"
 rsource "mcumgr_smp_client/Kconfig"
-
+rsource "nrf70_fw_ext/Kconfig"
 endmenu
```
Comment (nordicjm) [inline subsys/net/lib/Kconfig:51]: +newline

File: subsys/net/lib/nrf70_fw_ext/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#
+# Copyright (c) 2023 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+if (not CONFIG_NRF_WIFI_PATCHES_BUILTIN)
```
Comment (nordicjm) [inline subsys/net/lib/nrf70_fw_ext/CMakeLists.txt:7]:
```suggestion
if(NOT CONFIG_NRF_WIFI_PATCHES_BUILTIN)
```

File: subsys/net/lib/nrf70_fw_ext/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+zephyr_library_sources_ifdef(ext_fw_load.c)
+if(CONFIG_NRF_WIFI_PATCHES_EXT_FLASH_XIP)
+# Run patches from the external flash (XIP). No need to copy.
+zephyr_code_relocate(FILES ext_fw_load.c LOCATION EXTFLASH_RODATA NOCOPY)
```
Comment (nordicjm) [inline subsys/net/lib/nrf70_fw_ext/CMakeLists.txt:11]: indent

File: subsys/net/lib/nrf70_fw_ext/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+endif()
```
Comment (nordicjm) [inline subsys/net/lib/nrf70_fw_ext/CMakeLists.txt:12]: this cannot have been tested because it is missing an endif

File: sysbuild/CMakeLists.txt [hunk 5]
```diff
@@ -427,10 +427,15 @@ function(${SYSBUILD_CURRENT_MODULE_NAME}_pre_cmake)
     foreach(config WIFI_PATCHES_EXT_FLASH_DISABLED WIFI_PATCHES_EXT_FLASH_XIP
             WIFI_PATCHES_EXT_FLASH_STORE)
       if(${SB_CONFIG_${config}})
+        message("Setting ${config} from SB")
```
Comment (nordicjm) [inline sysbuild/CMakeLists.txt:430]: ?

File: sysbuild/CMakeLists.txt [hunk 6]
```diff
@@ -427,10 +427,15 @@ function(${SYSBUILD_CURRENT_MODULE_NAME}_pre_cmake)
...
         set_config_bool(${DEFAULT_IMAGE} CONFIG_NRF_${config} ${SB_CONFIG_${config}})
       endif()
     endforeach()
 
+    if (NOT DEFINED SB_CONFIG_WIFI_PATCHES_EXT_FLASH_DISABLED)
```
Comment (nordicjm) [inline sysbuild/CMakeLists.txt:435]:
```suggestion
    if(NOT DEFINED SB_CONFIG_WIFI_PATCHES_EXT_FLASH_DISABLED)
```

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: subsys/net/lib/nrf70_fw_ext/CMakeLists.txt (hunk 4 above)
Comment (krish2718) [reply subsys/net/lib/nrf70_fw_ext/CMakeLists.txt:12]: I did only test build for XIP, this might have creptin with some rebase later.

**Review by sachinthegreen — APPROVED** [review]

File: subsys/net/lib/nrf70_fw_ext/ext_fw_load.c [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+#else
+LOG_MODULE_DECLARE(wifi_nrf, CONFIG_WIFI_NRF70_LOG_LEVEL);
+#endif /* CONFIG_NRF_WIFI_PATCHES_EXT_FLASH_XIP */
+
+#include <fmac_main.h>
+
+#ifdef CONFIG_NRF_WIFI_PATCHES_EXT_FLASH_XIP
+/* INCBIN macro Taken from https://gist.github.com/mmozeiko/ed9655cf50341553d282 */
+#define STR2(x) #x
+#define STR(x) STR2(x)
```
Comment (rado17) [inline subsys/net/lib/nrf70_fw_ext/ext_fw_load.c:36]: Just for my understanding, why do we need `STR2` when same `stringify` operation can be done with `STR`?

**Review by rado17 — APPROVED** [review]

File: subsys/net/lib/nrf70_fw_ext/ext_fw_load.c (hunk 7 above)
Comment (krish2718) [reply subsys/net/lib/nrf70_fw_ext/ext_fw_load.c:36]: First one converts to a string, second expands it's value.

Comment (gmarull) [inline subsys/net/lib/nrf70_fw_ext/ext_fw_load.c:36]: use `STRINGIFY` from Zephyr

**Review by gmarull — DISMISSED** [review]

File: subsys/net/lib/nrf70_fw_ext/ext_fw_load.c (hunk 7 above)
Comment (krish2718) [reply subsys/net/lib/nrf70_fw_ext/ext_fw_load.c:36]: Just had a chat with Jamie while trying to fix another problem, will remove INCBIN and use `generate_inc_file_for_target` to add proper dependencies.

File: samples/matter/lock/boards/nrf7002dk_nrf5340_cpuapp.conf [hunk 8]
```diff
@@ -0,0 +1,9 @@
+#
+# Copyright (c) 2022 Nordic Semiconductor ASA
```
Comment (maciejbaczmanski) [inline samples/matter/lock/boards/nrf7002dk_nrf5340_cpuapp.conf:2]:
```suggestion
# Copyright (c) 2024 Nordic Semiconductor ASA
```

**Comment by krish2718** [comment]
@rlubos  for two of the twister failures, this https://github.com/nrfconnect/sdk-zephyr/pull/2072/commits/4805f6304893fe145bdde2e8266bbc4ebde76784 commit is needed, how should we proceed? But looks like that PR takes time to merge, should I just fix the Wi-Fi sample only in this PR?

**Comment by krish2718** [comment]
> @rlubos for two of the twister failures, this [nrfconnect/sdk-zephyr@4805f63](https://github.com/nrfconnect/sdk-zephyr/commit/4805f6304893fe145bdde2e8266bbc4ebde76784) commit is needed, how should we proceed? But looks like that PR takes time to merge, should I just fix the Wi-Fi sample only in this PR?

FYI https://github.com/nrfconnect/sdk-zephyr/pull/2062/commits/77dd806bc563040b77158d209f9e07845eb7987c for CI to proceed, we can decide once CI is green.

**Comment by rlubos** [comment]
@krish2718 My PR https://github.com/nrfconnect/sdk-zephyr/pull/2072 is practically ready, I'm not merging it yet intentionally because we have another large set of cherry-pick which has top priority now (https://github.com/nrfconnect/sdk-nrf/pull/17721) and don't want disrupt that process. Will merge as soon as we'll done with that one.

File: subsys/net/lib/nrf70_fw_ext/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,13 @@
+#
+# Copyright (c) 2023 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+if (NOT CONFIG_NRF_WIFI_PATCHES_BUILTIN)
```
Comment (nordicjm) [inline subsys/net/lib/nrf70_fw_ext/CMakeLists.txt:7]:
```suggestion
if(NOT CONFIG_NRF_WIFI_PATCHES_BUILTIN)
```

File: subsys/net/lib/nrf70_fw_ext/ext_fw_load.c [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+LOG_MODULE_DECLARE(wifi_nrf, LOG_LEVEL_NONE);
+#else
+LOG_MODULE_DECLARE(wifi_nrf, CONFIG_WIFI_NRF70_LOG_LEVEL);
+#endif /* CONFIG_NRF_WIFI_PATCHES_EXT_FLASH_XIP */
+
+#include <fmac_main.h>
+
+#ifdef CONFIG_NRF_WIFI_PATCHES_EXT_FLASH_XIP
+static const char nrf70_fw_patch[] = {
+	#include <nrf70_fw_patch/nrf70.bin.inc>
```
Comment (nordicjm) [inline subsys/net/lib/nrf70_fw_ext/ext_fw_load.c:35]: where is this file generated?

File: cmake/sysbuild/nrf700x.cmake [hunk 11]
```diff
@@ -8,16 +8,16 @@ function(setup_nrf700x_xip_data)
   sysbuild_dt_nodelabel(qspi_nodelabel IMAGE ${DEFAULT_IMAGE} NODELABEL "qspi")
   sysbuild_dt_reg_addr(qspi_xip_address IMAGE ${DEFAULT_IMAGE} PATH "${qspi_nodelabel}" NAME "qspi_mm")
 
-  set(OS_AGNOSTIC_BASE ${ZEPHYR_NRFXLIB_MODULE_DIR}/nrf_wifi)
+  set(NRF70_FW_BINS ${CMAKE_CURRENT_LIST_DIR}/../../../modules/hal/nordic/zephyr/blobs/wifi_fw_bins/)
```
Comment (nordicjm) [inline cmake/sysbuild/nrf700x.cmake:11]: use correct variable

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: subsys/net/lib/nrf70_fw_ext/ext_fw_load.c (hunk 10 above)
Comment (krish2718) [reply subsys/net/lib/nrf70_fw_ext/ext_fw_load.c:35]: https://github.com/nrfconnect/sdk-zephyr/pull/2062/commits/30100c70239da1201849ec673fde8228a5cbdd98#diff-9171118f1e64973f46825a6d56db0770019da60848725b7256a131d045a3a706R175

**Review by rlubos — APPROVED** [review]

**Review by ArekBalysNordic — APPROVED** [review]

File: subsys/net/lib/nrf70_fw_ext/rpu_fw_patches.ld [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+#endif /* CONFIG_BOARD_NRF5340DK_NRF5340_CPUAPP */
+
+#if USE_PARTITION_MANAGER && PM_EXTERNAL_FLASH_ADDRESS
+#include <pm_config.h>
+#define EXTFLASH_ADDRESS (EXTFLASH_BASE_ADDR + PM_EXTERNAL_FLASH_ADDRESS)
+#undef EXTFLASH_SIZE
+#define EXTFLASH_SIZE (PM_EXTERNAL_FLASH_SIZE)
+#else
+#define EXTFLASH_ADDRESS (EXTFLASH_BASE_ADDR)
+#endif /* USE_PARTITION_MANAGER && PM_EXTERNAL_FLASH_ADDRESS */
```
Comment (nordicjm) [inline subsys/net/lib/nrf70_fw_ext/rpu_fw_patches.ld:42]: this should really have it's own PM section that is dedicated, right now it looks like it's reading all of the external qspi flash

Comment (krish2718) [reply subsys/net/lib/nrf70_fw_ext/rpu_fw_patches.ld:42]: Sure, will take it up separately in another PR as this was a old file.

File: subsys/net/lib/nrf70_fw_ext/Kconfig [hunk 13]
```diff
@@ -0,0 +1,96 @@
+#
+# Copyright (c) 2024 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+if WIFI_NRF70 && !NRF_WIFI_BUILD_ONLY_MODE
+choice NRF_WIFI_PATCHES_EXT_FLASH_SUPPORT
```
[the remark this answers is not in the document]
Comment (krish2718) [reply subsys/net/lib/nrf70_fw_ext/Kconfig:8]: will fix during the manifest update.

**Pull request merged**